File: Data Structures & Algorithms/topologicalSort/submission-1.py
class Solution:
    def topologicalSort(self, n: int, edges: List[List[int]]) -> List[int]:
        mp = defaultdict(list)

        for u, v in edges:
            mp[v].append(u)
        
        visited = [False for _ in range(n)]
        path = [False for _ in range(n)]
        res = []
        def dfs(i):
            if path[i]:
                return False
            if visited[i]:
                return True
            
            visited[i] = True
            path[i] = True
            for nei in mp[i]:
                if not dfs(nei):
                    return False
            
            path[i] = False
            res.append(i)
            return True
        
        for i in range(n):
            if not dfs(i):
                return []
        
        return res

File: Data Structures & Algorithms/matrixDFS/submission-1.cpp
class Solution 
{
public:
    void dfs(vector<vector<int>>& grid, int r, int c, int& m, int& n, int& res)
    {
        if (r == m - 1 && c == n - 1)
        {
            res++;
            return;
        }

        if (r >= m || c >= n || r < 0 || c < 0 || grid[r][c] == 1)
        {
            return;
        }

        grid[r][c] = 1;
        dfs(grid, r + 1, c, m, n, res);
        dfs(grid, r - 1, c, m, n, res);
        dfs(grid, r, c + 1, m, n, res);
        dfs(grid, r, c - 1, m, n, res);
        grid[r][c] = 0;
    }
    int countPaths(vector<vector<int>>& grid) 
    {
        int m = grid.size();
        int n = grid.front().size();
        if (grid[m - 1][n - 1] == 1) return 0;
        int res = 0;

        dfs(grid, 0, 0, m, n, res);

        return res;
    }
};
